RecoEgamma/EgammaPhotonProducers/src/PhotonProducerHGCal.cc

```diff
@@ -0,0 +1,501 @@
+    }
+    if (validVertex) vertexCollection = *(vertexHandle.product());
+  }
+
+  int iSC = 0;  // index in photon collection
+  // Loop over barrel and endcap SC collections and fill the  photon collection
+  if (validPhotonCoreHandle)
+    fillPhotonCollection(theEvent, theEventSetup, photonCoreHandle, topology,
+                         &barrelRecHits, &endcapRecHits, hcalTowersHandle,
+                         vertexCollection, outputPhotonCollection, iSC,
+                         sevLv.product());
+
+  // put the product in the event
+  edm::LogInfo("PhotonProducerHGCal")
+      << " Put in the event " << iSC << " Photon Candidates \n";
+  outputPhotonCollection_p->assign(outputPhotonCollection.begin(),
+                                   outputPhotonCollection.end());
+  theEvent.put(std::move(outputPhotonCollection_p), PhotonCollection_);
+}
+
+void PhotonProducerHGCal::fillPhotonCollection(
+    edm::Event& evt, edm::EventSetup const& es,
+    const edm::Handle<reco::PhotonCoreCollection>& photonCoreHandle,
+    const CaloTopology* topology, const EcalRecHitCollection* ecalBarrelHits,
+    const EcalRecHitCollection* ecalEndcapHits,
+    const edm::Handle<CaloTowerCollection>& hcalTowersHandle,
+    reco::VertexCollection& vertexCollection,
+    reco::PhotonCollection& outputPhotonCollection, int& iSC,
+    const EcalSeverityLevelAlgo* sevLv) {
+  const CaloGeometry* geometry = theCaloGeom_.product();
+  const CaloSubdetectorGeometry* subDetGeometry = 0;
+  const CaloSubdetectorGeometry* geometryES =
+      theCaloGeom_->getSubdetectorGeometry(DetId::Ecal, EcalPreshower);
+  const EcalRecHitCollection* hits = 0;
+  std::vector<double> preselCutValues;
+  float minR9 = 0;
+
+  std::vector<int> flags_, severitiesexcl_;
+
+  for (unsigned int lSC = 0; lSC < photonCoreHandle->size(); lSC++) {
+    reco::PhotonCoreRef coreRef(reco::PhotonCoreRef(photonCoreHandle, lSC));
+    reco::SuperClusterRef scRef = coreRef->superCluster();
+    iSC++;
+
+    bool isHGCal = (scRef->seed()->seed().det() == DetId::Forward) ||
+                   (scRef->seed()->seed().det() == DetId::Hcal);
+    int subdet = scRef->seed()->hitsAndFractions()[0].first.subdetId();
+    subDetGeometry = theCaloGeom_->getSubdetectorGeometry(DetId::Ecal, subdet);
+
+    if (!isHGCal) {
+      if (subdet == EcalBarrel) {
+        preselCutValues = preselCutValuesBarrel_;
+        minR9 = minR9Barrel_;
+        hits = ecalBarrelHits;
+        flags_ = flagsexclEB_;
+        severitiesexcl_ = severitiesexclEB_;
+      } else if (subdet == EcalEndcap) {
+        preselCutValues = preselCutValuesEndcap_;
+        minR9 = minR9Endcap_;
+        hits = ecalEndcapHits;
+        flags_ = flagsexclEE_;
+        severitiesexcl_ = severitiesexclEE_;
+      } else {
+        edm::LogWarning("") << "PhotonProducerHGCal: do not know if it is a barrel or "
+                               "endcap SuperCluster";
+      }
+    } else {
+      // Compute everything HGCal related here inside.
```

Comment:
Generally this level of code duplication is not acceptable. It has been allowed in limited, temporary cases for upgrade before, because of time/effort constraints. Those constraints don't seem to apply for this relatively limited change.